bokehjs/src/coffee/models/ranges/factor_range.ts

```diff
@@ -0,0 +1,233 @@
+
+export function map_one_level(factors: L1Factor[], padding: number, offset: number = 0): [L1Mapping, number] {
+  const mapping: {[key: string]: {value: number}} = {}
+
+  for (let i = 0; i < factors.length; i++) {
+    const factor = factors[i]
+    if (factor in mapping)
+      throw new Error(`duplicate factor or subfactor: ${factor}`)
+    else
+      mapping[factor] = {value: 0.5 + i*(1 + padding) + offset}
+  }
+
+  return [mapping, (factors.length - 1)*padding]
+}
+
+export function map_two_levels(factors: L2Factor[],
+                               outer_pad: number, factor_pad: number,
+                               offset: number = 0): [L2Mapping, string[], number] {
+  const mapping: L2Mapping = {}
+
+  const tops: {[key: string]: string[]} = {}
+  const tops_order: string[] = []
+  for (const [f0, f1] of factors) {
+    if (!(f0 in tops)) {
+      tops[f0] = []
+      tops_order.push(f0)
+    }
+    tops[f0].push(f1)
+  }
+
+  let suboffset = offset
+  let total_subpad = 0
+  for (const f0 of tops_order) {
+    const n = tops[f0].length
+    const [submap, subpad] = map_one_level(tops[f0], factor_pad, suboffset)
+    total_subpad += subpad
+    const subtot = sum(tops[f0].map((f1) => submap[f1].value))
+    mapping[f0] = {value: subtot/n, mapping: submap}
+    suboffset += n + outer_pad + subpad
+  }
+
+  return [mapping, tops_order, (tops_order.length-1)*outer_pad + total_subpad]
+}
+
+export function map_three_levels(factors: L3Factor[],
+                                 outer_pad: number, inner_pad: number, factor_pad: number,
+                                 offset: number = 0): [L3Mapping, string[], [string, string][], number] {
+  const mapping: L3Mapping = {}
+
+  const tops: {[key: string]: [string, string][]} = {}
+  const tops_order: string[] = []
+  for (const [f0, f1, f2] of factors) {
+    if (!(f0 in tops)) {
+      tops[f0] = []
+      tops_order.push(f0)
+    }
+    tops[f0].push([f1, f2])
+  }
+
+  const mids_order: [string, string][] = []
+
+  let suboffset = offset
+  let total_subpad = 0
+  for (const f0 of tops_order) {
+    const n = tops[f0].length
+    const [submap, submids_order, subpad] = map_two_levels(tops[f0], inner_pad, factor_pad, suboffset)
+    for (const f1 of submids_order)
+      mids_order.push([f0, f1])
+    total_subpad += subpad
+    const subtot = sum(tops[f0].map(([f1,]) => submap[f1].value))
+    mapping[f0] = {value: subtot/n, mapping: submap}
+    suboffset += n + outer_pad + subpad
+  }
+
+  return [mapping, tops_order, mids_order, (tops_order.length-1)*outer_pad + total_subpad]
+}
+
+export class FactorRange extends Range {
+
+  factors: Factor[]
+  factor_padding: number
+  subgroup_padding: number
+  group_padding: number
+  range_padding: number
+  range_padding_units: PaddingUnits
+  start: number
+  end: number
+  bounds: [number, number] | "auto"
+  min_interval: any // XXX: what's this
+  max_interval: any // XXX: what's this
```

Comment:
also numbers

Reply:
Yeah, I actually answered this to my self when rewriting another type of range. I will have to revisit this anyway, because inheritance needs to be set up properly for these and several other properties.